### PR TITLE
feat(udf): support UDAF for Python and JS

### PR DESCRIPTION
**Pull request by wangrunji0408**
Signed-off-by: Runji Wang <wangrunji0408@163.com>I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

This PR adds aggregate function support for embedded Python and JavaScript UDF. The user interface is stateful. It looks like:

```sql
create aggregate my_sum(value int) returns int language python as $$
def create_state():
    return 0

def accumulate(state, value):
    return state + value

def retract(state, value):
    return state - value

def finish(state):
    return state
$$;

select my_sum(v) from generate_series(1, 5) as t(v);
----
15

drop aggregate my_sum;
```

User can define state as any type. The state will be serialized (using pickle in Python and json in JS) and then stored in our state table as a BYTEA value.

Change details are as follows:
- parser and frontend: support `create aggregate` and `drop aggregate`. the logic is basically the same as `create function`.
- frontend: add UDAF lookup when binding function. the code is a bit messy and may need to be refactored in the future.
- proto: introduce `AggKind::UserDefined` and add additional UDF metadata to `AggCall`.
- expr_core: add a new kind of aggregate `UserDefinedAggregateFunction`. add new methods to UDF interface. 
- expr_impl: upgrade `arrow_udf_python` and `arrow_udf_js` to include aggregate function support.
  - https://github.com/risingwavelabs/arrow-udf/pull/23
  - https://github.com/risingwavelabs/arrow-udf/pull/26

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [x] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

Support user-defined aggregate function in embedded Python and JavaScript UDF.

### CODE REVIEW

File: src/common/src/buffer/bitmap.rs [hunk 1]
```diff
@@ -763,6 +803,16 @@ impl<'a> iter::Iterator for BitmapOnesIter<'a> {
             }
         }
     }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        match self {
+            BitmapOnesIter::Range { start, end } => {
```
Comment (BugenZhao) [inline src/common/src/buffer/bitmap.rs:809]: Is this only used in all-zero and all-one cases?

File: src/expr/impl/src/udf/quickjs.rs [hunk 2]
```diff
@@ -72,4 +85,22 @@ impl UdfImpl for QuickJsFunction {
             .call_table_function(&self.identifier, input, 1024)
             .map(|s| futures_util::stream::iter(s).boxed())
     }
+
+    fn create_state(&self) -> Result<ArrayRef> {
+        self.runtime.create_state(&self.identifier)
```
Comment (BugenZhao) [inline src/expr/impl/src/udf/quickjs.rs:90]:
```suggestion
    fn agg_create_state(&self) -> Result<ArrayRef> {
        self.runtime.call_agg_create_state(&self.identifier)
```

What about making the signatures more clear?

**Review by BugenZhao — APPROVED** [review]
LGTM!!!

**Review by BugenZhao — APPROVED** [review]
LGTM!!!

**Comment by BugenZhao** [comment]
> using pickle in Python

By using pickle, does it mean that we should be more careful when bumping the version of preluded dependencies for the runtime?

**Comment by wangrunji0408** [comment]
> > using pickle in Python
> 
> By using pickle, does it mean that we should be more careful when bumping the version of preluded dependencies for the runtime? Not sure if the compatibility for these external dependencies can be preserved.

What "dependencies" do you refer to? I don't think compatibility is a problem since the Python code can not be changed after creation. 🤔

File: src/common/src/buffer/bitmap.rs (hunk 1 above)
Comment (wangrunji0408) [reply src/common/src/buffer/bitmap.rs:809]: Yes. I added this function because arrow `BooleanArray::from_iter` requires a upper bound of the iterator.

**Comment by BugenZhao** [comment]
> What "dependencies" do you refer to?

For example, do we provision dependencies such as  `pandas`? If so, will we bump them regularly?

The Python version won't be a problem. I'm just considering whether the internal representation of the objects from external dependencies will change.

**Comment by wangrunji0408** [comment]
Oh, we don't allow third-party libraries like pandas yet. Only a very limited set of builtin libraries are allowed for now. If we open the gate for third-party libraries in the future, we may record the exact version of each dependency (like cargo.lock).

Anyway, this is a good question. 👍

**Pull request merged**